core/state/snapshot/generate.go

```diff
@@ -40,17 +43,67 @@ var (
+	// is running. In this case the generation is aborted and wait the new signal.
+	errMissingTrie = errors.New("missing trie")
+)
+
+// Metrics in generation
+var (
+	snapGeneratedAccountMeter     = metrics.NewRegisteredMeter("state/snapshot/generation/account/generated", nil)
+	snapRecoveredAccountMeter     = metrics.NewRegisteredMeter("state/snapshot/generation/account/recovered", nil)
+	snapWipedAccountMeter         = metrics.NewRegisteredMeter("state/snapshot/generation/account/wiped", nil)
+	snapMissallAccountMeter       = metrics.NewRegisteredMeter("state/snapshot/generation/account/missall", nil)
+	snapGeneratedStorageMeter     = metrics.NewRegisteredMeter("state/snapshot/generation/storage/generated", nil)
+	snapRecoveredStorageMeter     = metrics.NewRegisteredMeter("state/snapshot/generation/storage/recovered", nil)
+	snapWipedStorageMeter         = metrics.NewRegisteredMeter("state/snapshot/generation/storage/wiped", nil)
+	snapMissallStorageMeter       = metrics.NewRegisteredMeter("state/snapshot/generation/storage/missall", nil)
+	snapSuccessfulRangeProofMeter = metrics.NewRegisteredMeter("state/snapshot/generation/proof/success", nil)
+	snapFailedRangeProofMeter     = metrics.NewRegisteredMeter("state/snapshot/generation/proof/failure", nil)
+
+	snapAccountProveTimer    = metrics.NewRegisteredGauge("state/snapshot/generation/duration/account/prove", nil)
+	snapAccountTrieReadTimer = metrics.NewRegisteredGauge("state/snapshot/generation/duration/account/trieread", nil)
+	snapAccountSnapReadTimer = metrics.NewRegisteredGauge("state/snapshot/generation/duration/account/snapread", nil)
+	snapAccountWriteTimer    = metrics.NewRegisteredGauge("state/snapshot/generation/duration/account/write", nil)
+	snapStorageProveTimer    = metrics.NewRegisteredGauge("state/snapshot/generation/duration/storage/prove", nil)
+	snapStorageTrieReadTimer = metrics.NewRegisteredGauge("state/snapshot/generation/duration/storage/trieread", nil)
+	snapStorageSnapReadTimer = metrics.NewRegisteredGauge("state/snapshot/generation/duration/storage/snapread", nil)
+	snapStorageWriteTimer    = metrics.NewRegisteredGauge("state/snapshot/generation/duration/storage/write", nil)
+)
+
+// Global timer for metrics
+var (
+	accountProving  time.Duration // The total time spent on the account proving
+	accountTrieRead time.Duration // The total time spent on the account trie iteration
+	accountSnapRead time.Duration // The total time spent on the snapshot account iteration
+	accountWrite    time.Duration // The total time spent on writing/updating/deleting accounts
+	storageProving  time.Duration // The total time spent on the storage proving
+	storageTrieRead time.Duration // The total time spent on the storage trie iteration
+	storageSnapRead time.Duration // The total time spent on the snapshot storage iteration
+	storageWrite    time.Duration // The total time spent on writing/updating/deleting storages
```

Comment:
I don't think these are a good idea.

First, I think Gauges have an `Inc` method that you could use to bump the counters up without having to track the totals at all. Still, I think tracking them as a single number is a bad idea.

If you were to use `Timers` (or `ResettingTimers`, even better) instead of the Gauges, you could see the evolution of the times across different ranges of the chain, and AFAIK timers also have a `count` method that maintains the total, so you'd still have the possibility to retrieve the absolute final sum like you do with gauges.

Reply:
The count in the `Timers` or `ResettingTimers` is used for counting total event number, but not for the sum for all events(duration here).

Reply:
Oh, sorry.. I just pushed a commit to use actual Timer meters. Feel free to drop my commit

Reply:
I've changed the timers to counters. There's no need to have these though, we can just increment the counter (could have done with the gauges too) instead of tracking outside and setting the exact value on update.